### ISSUE TITLE
[Question] Light & dark available, light by default

### ISSUE THREAD

**Issue by KaKi87** [darkreader/darkreader#1837]
Hello,

_It is very likely that the question has been answered many times already, and I tried searching in issues, but I'm sorry I can't find it._

_Therefore, I'd like to suggest adding this question to the README._

> If some website is **already dark** (has all pages dark by default), you can **add it to [dark-sites.config](https://github.com/alexanderby/darkreader/blob/master/src/config/dark-sites.config) file**
*(please, preserve alphabetical order)*.

The question is : what if a website has a light theme by default, but offers a dark mode which requires to click on some toggle to enable it ?

I think that a specific list cloud be created for this, then, a notice could be displayed under the switch to indicate that the website has a built-in dark mode, and, if possible, enable it automatically, as long as it is possible to detect that it is not already enabled.

Thanks

**Comment by KaKi87** [comment]
up

**Comment by WebMechanic** [comment]
> as long as it is possible to detect that it is not already enabled

web designers have a myriad of possibilities to enable and switch between themes and/or styles. 
There's no way to detect which method - or combination thereof - is currently provided. And website themes change occasionally after a relaunch/redesign.
Some also add goodies like time or season specific themes or other variations for a11y purposes.

**Comment by KaKi87** [comment]
I would imagine this feature that way :

- if the theme option is stored server-side, there is some API endpoint to call
- if the theme option is stored client-side, there is some local storage or cookie value to edit

We would just have to store that information.

Am I wrong ?

**Comment by WebMechanic** [comment]
I don't think you could convince *any* website owner to support some arbitrary "API endpoint" (that would need a proper and *stable* specification to begin with) for some arbitrary browser extension no matter it's popularity. Any implementation usually costs money after all.
Unfortunately #a11y of any kind has a low priority for the average website owner.

> some local storage or cookie

**TL;DR** no, you're not wrong for the second part :)

That's more likely to work and feasible.
I created client side "theme switchers" or "font scalers" myself and saved settings exactly like that and browser extensions have their own data storage anyway.

I'm unfamiliar with DM's inner workings though. 
If it can/is allowed to "record" user interaction with a webpage (listening to events, DOM changes) it might record the steps/clicks a user performs to enable such a theme switcher; ie. pressed element with id xx / changed value of checkbox/radio button yy / created cookie zz". 
Like a **macro recorder**. 
That information should then be enough to replay back. There are several extensions that do client side automation. I dunno if that would require additional access rights for DW. 

As there still might be some server side component involved for any site, timing may become an issue: like **when** would this action be played back to avoid flickering or FOUC? 
Loading a page triggers various events and happens in stages and any site may need a different "entry point" for the theme to switch. However, that too, might become part of some DM setting for this.

Just brainstorming my 2ct >=)

**Comment by KaKi87** [comment]
Sorry, but you misunderstood my point.

What I meant is that just like Dark Reader stores specific CSS rules for specific websites in its configuration, it could store the specificatins of the light/dark toggling method for each website which has its own switch.

For example :

- On betaseries.com, the endpoint is :

```js
GET `https://www.betaseries.com/ajax/account/option.php?name=tm_site&value=${value}`
```

Where `value` equals `1` for *dark* and `0` for *light*.

<br>

- On reddit.com, the endpoint is :

```js
PATCH `https://oauth.reddit.com/api/v1/me/prefs?raw_json=1&gilding_detail=1`
```

With payload :

```js
`{ nightmode: ${value} }`
```

Where `value` equals `true` for *dark* and `false` for *light*.

Etc...

**Comment by WebMechanic** [comment]
ah, ok. got it!
So for an already existing server side switch you'd just store the URL and parameter that toggle the theme. That could work if that's all it takes. 
Could be fiddly though if it needs some valid session data or request tokens to go with it in the HTTP header.

Cookies and localStorage are bound to a domain. If a dark/light theme state is stored there, IIRC extensions would not be allowed to access that information. If DR stores the state as well those settings might get out of sync.

**Comment by KaKi87** [comment]
Actually, I read that you can access cookies using the WebExtensions API and localStorage directly from the background script.

**Comment by FranklinYu** [comment]
#1327

**Comment by KaKi87** [comment]
Thanks @FranklinYu, I'm closing this issue since it's a duplicate.

**Comment by KaKi87** [comment]
Yet another example : on regex101.com, the theme setting is in :

```js
JSON.parse(localStorage['regex101-state']).settings.theme
```

I'll continue to post some stuff like that here just for saving time, if no one minds about it.

**Comment by KaKi87** [comment]
On nextinpact.com, the endpoint is :

```js
POST `https://www.nextinpact.com/switch-design`
```

With payload :

```js
`{ designId: ${value} }`
```

Where `value` equals `2` for *dark* and `1` for *light*.

**Comment by KaKi87** [comment]
On stackoverflow.com, the endpoint is :

```js
POST `https://stackoverflow.com/account/set-theme-preference`
```

With payload :

```js
`{ theme: ${value} }`
```

Where `value` equals `1` for *dark* and `0` for *light*. (And `2` for *system*)

**Comment by KaKi87** [comment]
On deezer.com, the endpoint is :

```js
POST `https://www.deezer.com/ajax/gw-light.php?method=user.setSettings&input=3&api_version=1.0&api_token=${api_token}&cid=${cid}`
```

With query params : `api_token` & `cid` ; Location unknown yet.

With payload :

```js
`{ settings: { global: dark_mode: ${value} } }`
```

Where `value` equals `on` for *dark* and `off` for *light*.
